cmd/prow-job-dispatcher/main.go

```diff
@@ -61,9 +83,244 @@ func (o *options) complete(secrets *sets.String) error {
+			}
+		}
+	}
+	for _, job := range jc.Periodics {
+		if ct := getClusterTypeFromEnv(job.Spec); ct != "" {
+			cloudProviders.Insert(ct)
+		}
+	}
+	return cloudProviders
+}
+
+type clusterVolume struct {
+	// [cloudProvider][cluster]volume
+	clusterVolumeMap map[string]map[string]float64
+	mutex            sync.Mutex
+}
+
+func (cv *clusterVolume) dispatchToCluster(cloudProvider string, jc *prowconfig.JobConfig, path string, config *dispatcher.Config, jobVolumes map[string]float64) string {
+	//no cluster in the build farm is from the targeting cloud provider
+	if _, ok := cv.clusterVolumeMap[cloudProvider]; !ok {
+		cloudProvider = ""
+	}
+	var cluster, rCloudProvider string
+	min := float64(-1)
+	cv.mutex.Lock()
+	for cp, m := range cv.clusterVolumeMap {
+		for c, v := range m {
+			if cloudProvider == "" || cloudProvider == cp {
+				if min < 0 || min > v {
+					min = v
+					cluster = c
+					rCloudProvider = cp
+				}
+			}
+		}
+	}
+
+	var volume float64
+	for k := range jc.PresubmitsStatic {
+		for _, job := range jc.PresubmitsStatic[k] {
+			cluster1, isForBuildFarm := config.DetermineClusterForJob(job.JobBase, path)
+			if cluster == cluster1 || isForBuildFarm {
+				volume = volume + jobVolumes[job.JobBase.Name]
+			}
+		}
+	}
+
+	for k := range jc.PostsubmitsStatic {
+		for _, job := range jc.PostsubmitsStatic[k] {
+			cluster1, isForBuildFarm := config.DetermineClusterForJob(job.JobBase, path)
+			if cluster == cluster1 || isForBuildFarm {
+				volume = volume + jobVolumes[job.JobBase.Name]
+			}
+		}
+	}
+	for _, job := range jc.Periodics {
+		cluster1, isForBuildFarm := config.DetermineClusterForJob(job.JobBase, path)
+		if cluster == cluster1 || isForBuildFarm {
+			volume = volume + jobVolumes[job.JobBase.Name]
+		}
+	}
+
+	if cv.clusterVolumeMap[rCloudProvider] == nil {
+		cv.clusterVolumeMap[rCloudProvider] = map[string]float64{}
+	}
+	cv.clusterVolumeMap[rCloudProvider][cluster] = cv.clusterVolumeMap[rCloudProvider][cluster] + volume
+
+	cv.mutex.Unlock()
+	return cluster
+}
+
+// dispatchJobConfig dispatches the jobs defined in a Prow jon config
+func dispatchJobConfig(jc *prowconfig.JobConfig, path string, config *dispatcher.Config, jobVolumes map[string]float64, cv *clusterVolume) string {
+	cloudProvidersForE2ETests := getCloudProvidersForE2ETests(jc)
+	var cluster string
+	if cloudProvidersForE2ETests.Len() == 1 {
+		cloudProvider, _ := cloudProvidersForE2ETests.PopAny()
+		cluster = cv.dispatchToCluster(cloudProvider, jc, path, config, jobVolumes)
+	} else {
+		cluster = cv.dispatchToCluster("", jc, path, config, jobVolumes)
+	}
+
+	return cluster
+}
+
+type configResult struct {
+	cluster  string
+	filename string
+}
+
+// dispatchJobs loads the Prow jobs and chooses a cluster in the build farm if possible.
+// The current implementation walks through the Prow Job config files.
+// For each file, it tries to assign all jobs in it to a cluster in the build farm: dispatchJobConfig(jobConfig, path, config, jobVolumes, cv)
```

Comment:
This is still way too vague, sorry. Please do something like:
* It always assigns all jobs in one file to the same cluster
* It first checks if all jobs use the same cloud provider and then will use a cluster on that cloud provider
* If there is more than one cluster on that cloudprovider, it will choose the first
* If there are jobs for multiple cloud provider it does some other fancy stuff

Reply:
Updated (not exactly the same as suggested though).